language/src/typechecker.rs

```diff
@@ -975,9 +1018,27 @@ fn typecheck_expression(
                             arm_exp,
                             func_return_type,
                             top_level_context,
-                            &intermediate_var_context.clone().union(new_var_context),
+                            &VarContext {
+                                vars: intermediate_var_context
+                                    .clone()
+                                    .vars
+                                    .union(new_var_context.vars),
+                                mutable_vars: intermediate_var_context
+                                    .clone()
+                                    .mutable_vars
+                                    .union(new_var_context.mutable_vars),
+                            },
```

Comment:
Maybe we could have a `merge` method on `VarContext`, that takes a `enum MergeKind {UnionVars, IntersectVars}` and an other `VarContext`?
That would factorize the code, but also make it easier to read, I think that's quite error-prone currently.